File: scrimage-io/src/main/scala/com/sksamuel/scrimage/nio/TwelveMonkeysWriter.scala
package com.sksamuel.scrimage.nio

import java.io.OutputStream
import javax.imageio.{ IIOImage, ImageIO }

import com.sksamuel.scrimage.Image
import org.apache.commons.io.IOUtils

trait TwelveMonkeysWriter extends ImageWriter {

  def format: String

  override def write(image: Image, out: OutputStream): Unit = {

    import scala.collection.JavaConverters._

    val writers = ImageIO.getImageWritersByFormatName(format).asScala.toList
    writers match {
      case writer :: tail =>
        val ios = ImageIO.createImageOutputStream(out)
        val params = writer.getDefaultWriteParam()
        writer.setOutput(ios)
        writer.write(null, new IIOImage(image.awt, null, null), params)
        ios.close()
        writer.dispose()
        IOUtils.closeQuietly(out)
      case _ => throw new UnsupportedOperationException(s"$format format not registered")
    }
  }
}

File: scrimage-io-extra/src/main/java/com/sksamuel/scrimage/nio/TwelveMonkeysWriter.java
package com.sksamuel.scrimage.nio;

import com.sksamuel.scrimage.Image;
import com.sksamuel.scrimage.ImageWriter;
import org.apache.commons.io.IOUtils;

import javax.imageio.IIOImage;
import javax.imageio.ImageIO;
import javax.imageio.ImageWriteParam;
import javax.imageio.stream.ImageOutputStream;
import java.io.IOException;
import java.io.OutputStream;

abstract class TwelveMonkeysWriter implements ImageWriter {

    public abstract String format();

    @Override
    public void write(Image image, OutputStream out) throws IOException {
        javax.imageio.ImageWriter writer = ImageIO.getImageWritersByFormatName(format()).next();
        ImageOutputStream ios = ImageIO.createImageOutputStream(out);
        ImageWriteParam params = writer.getDefaultWriteParam();
        writer.setOutput(ios);
        writer.write(null, new IIOImage(image.awt(), null, null), params);
        ios.close();
        writer.dispose();
        IOUtils.closeQuietly(out);
    }
}